go/src/koding/klient/machine/mount/anteroom.go

```diff
@@ -0,0 +1,277 @@
+		if !atomic.CompareAndSwapUint64((*uint64)(&ev.stat), uint64(statusPop), uint64(statusDeprecated)) {
+			return ev.ctx
+		}
+
+		// Change is deprecated. Re-push new change to the queue but keep
+		// context from old event.
+		newEv := NewEventCopy(ev)
+		a.evs[c.Name()] = newEv
+		a.queue.Push(newEv)
+		a.wakeup()
+	}
+
+	return ev.ctx
+}
+
+// Events is a dispatcher for the queued events. The returned channel will be
+// closed when Anteroom object is closed.
+func (a *Anteroom) Events() <-chan *Event {
+	return a.evC
+}
+
+// Status reports the current status of Anteroom object. The items value can
+// be interpreted as a number of files waiting for synchronization. Comparing
+// items and queued events shows how fast syncers are able to synchronize files.
+func (a *Anteroom) Status() (items int, queued int) {
+	a.mu.Lock()
+	items = len(a.evs)
+	a.mu.Unlock()
+
+	return items, a.queue.Size()
+}
+
+// Close stops the dynamic client. After this function is called, client is
+// in disconnected state and each contexts returned by it are closed.
+func (a *Anteroom) Close() {
+	a.once.Do(func() {
+		a.mu.Lock()
+		defer a.mu.Unlock()
+
+		// Mark all events as deprecated and detach them from the queue.
+		for name, ev := range a.evs {
+			atomic.StoreUint64((*uint64)(&ev.stat), uint64(statusDeprecated))
+			ev.cancel()
+
+			delete(a.evs, name)
+		}
+
+		a.closed = true
+		close(a.stopC) // Stop dispatching go-routine.
+	})
+}
+
+// dequeue pops events from the queue and sends them to events channel.
+func (a *Anteroom) dequeue() {
+	var (
+		ev  *Event      = nil
+		evC chan *Event = nil
+
+		// wakeupTick ensures that dequeue will be always waked up. Even in case
+		// we somehow miss the wakeup event. This is a safety fall-back.
+		wakeupTick = time.NewTicker(30 * time.Second)
+	)
+
+	tryPop := func() {
+		// Event haven't been sent yet.
+		if ev != nil {
+			return
+		}
+
+		if ev = a.queue.Pop(); ev == nil {
+			evC = nil // queue is empty - turn off event channel.
+		} else {
+			atomic.StoreUint64((*uint64)(&ev.stat), uint64(statusPop))
+			evC = a.evC // there is an event - turn on event channel.
+		}
+	}
+
+	for {
+		select {
+		case evC <- ev:
+			if ev = a.queue.Pop(); ev == nil {
+				evC = nil // ueue is empty - turn off event channel.
+			} else {
+				atomic.StoreUint64((*uint64)(&ev.stat), uint64(statusPop))
+			}
+		case <-a.wakeupC:
+			tryPop()
+		case <-wakeupTick.C:
+			tryPop()
+		case <-a.stopC:
+			wakeupTick.Stop()
+			close(a.evC)
+			return
+		}
+	}
+}
+
+// wakeup is a non-blocking attempt to wake up dequeue go-routine. Wakeup
+// channel is buffered and no more than one wakeup event is needed to wakeup
+// the dequeue loop so it's safe to drop excessive calls.
+func (a *Anteroom) wakeup() {
+	select {
+	case a.wakeupC <- struct{}{}:
+	default:
+	}
+}
+
+// detach removes the change from coalescing events map only if stored id
+// is equal to provided one.
+func (a *Anteroom) detach(name string, id uint64) {
```

Comment:
can we use `path` instead of `name`, it's easier to relate the login to other packages